### ISSUE TITLE
UI issues

### ISSUE THREAD

**Issue by gvcormac** [trec-dd/trec-dd-assessing#10]
-- render raw HTML is unsafe
-- if I accidentally follow a link in the raw html, there is no indication that I have left the review platform
-- it is almost impossible to highlight a live link without following it
-- no right-to-left scroll; can't read wide pages
-- don't know how to return to search and/or browse screen 
-- if you drag a passage and don't click a button, it is judged "not relevant"
-- should be able to refresh waterloo ranked list without judging ALL 20 docs.

**Comment by shawn67** [comment]
1. I think I need clear instruction on how to display the crawled web pages, I know it's not safe and just keep it there temporary. Before I just keep it display raw HTML code, which Grace think is unreadable.
2. Now the refreshing approach is when the remaining un-judged docs are less than a threshold, the sys will fetch new recommending list.

**Comment by gvcormac** [comment]
Regarding unsafe HTML:
1.  Ian, can you comment on what the platforms used to assess previous Web Tracks did, with regard to this issue, and in general?
2.  Waterloo has explored various options:

2(a) Render the HTML as text.  [Then display with &lt;pre>, being sure to escape "&lt;" to "&amp;lt;" and "&amp;" to "&amp;amp;"]  We have rendered the 1/2 million blackhat and hackforum documents, and the 1/2 million Ebola documents as text.  We could easily make these renderings available.

2(b) Renter HTML as pdf.  [This will take a long time, and requires a pdf viewer plugin.]

2(c) Use the "noscript" Firefox plugin.  [This is what we did for assessing Web Spam.  It helps, but is
somewhat intrusive]
1.  The problem in the Georgetown UI is exacerbated by the fact that there are no external controls.  If the panel containing the rendering gets hijacked, you can no longer navigate.  I could not even find a way to start over, as logging out and in simply restored the broken state.

**Comment by johnrfrank** [comment]
TREC KBA also did this type of assessing. And we spoke with Web Track and implemented the same kind of cleansing strategy that they used.   You can see the details in streamcorpus-pipeline. 

The key pieces are to use something like lxml.html.clean.Cleanser to remove all of the JavaScript and other pieces of non-visible content but leave behind the primary elements of formatting. 

The second piece is to generate clean visible, by replacing all of the tags with whitespace of the same length as the tag string. As a quick and dirty solution you could do the same thing here without using a cleanser like the one from lxml. That might make ugly content, and it will certainly remove any dangerous  JavaScrip.

**Comment by gvcormac** [comment]
Silly me.  I did not know that Github renders comments as raw html, so my comment regarding html rendering was messed up.  I have now fixed it.  Please re-read.

**Comment by isoboroff** [comment]
John's comment is spot on.  I highly recommend the lxml cleanser, Shiqi, as it's right there in Python for you.  In the web track we did it by hand but these days I think the tooling's gotten better.

That should be easy to put in there.  Then we can look at the intermediate result before working on clean-visible rendering... cleansed HTML should at least not fight with the browser so much.

**Comment by isoboroff** [comment]
Rendering is a duplicate of issue #1 by the way,

**Comment by johnrfrank** [comment]
> John's comment is spot on. I highly recommend the lxml cleanser, Shiqi, 
> as it's right there in Python for you. In the web track we did it by 
> hand but these days I think the tooling's gotten better.
> 
> That should be easy to put in there. Then we can look at the 
> intermediate result before working on clean-visible rendering... 
> cleansed HTML should at least not fight with the browser so much.

These are the two steps that we used in TREC KBA.  They've run several 
billion times now, so the dressing around them is reasonably hardened:

https://github.com/trec-kba/streamcorpus-pipeline/blob/master/streamcorpus_pipeline/_clean_html.py#L74

https://github.com/trec-kba/streamcorpus-pipeline/blob/master/streamcorpus_pipeline/_clean_visible.py#L75

jrf

**Comment by isoboroff** [comment]
Yes, Shiqi should be able to lift that code and stick it pretty much verbatim into the rendering path.

**Comment by JiyunLuo** [comment]
@johnrfrank I tried to use lxml.html.clean to remove html tags. It seems that some html tags are removed but many are still remained. Is this correct by using lxml.html.clean - not all tags will be removed? 

This is my configuration for lxml.html.clean:
    cleaner = lxml.html.clean.Cleaner(
        scripts=True, javascript=True,
        comments=True,
        # do not remove <html> <head> <title> etc
        page_structure=False,
        meta=True,
        style=True, links=True)

**Comment by johnrfrank** [comment]
> @johnrfrank I tried to use lxml.html.clean to remove html tags. It seems 
> that some html tags are removed but many are still remained. Is this 
> correct by using lxml.html.clean - not all tags will be removed?

I see why the name "clean html" is confusing.  It does not _remove_ tags, 
it simply takes a raw html string and cleans it to have correct HTML tags, 
and it can remove things that you don't want such as CSS and JavaScript. 
The result of cleansing is a "good" string of HTML, which can be displayed 
safely in a UI.  It retains HTML that may make the result more readable.

If you want to fully strip tags, you need to do something like the 
clean_visible transform linked earlier.

jrf

**Comment by JiyunLuo** [comment]
@johnrfrank 
Dear John,

Do you have an example about how to call these two python scripts?

Best,
Jiyun

**Comment by johnrfrank** [comment]
This example shows how.  It is two lines:

https://github.com/dossier/dossier.models/blob/master/dossier/models/etl.py#L59
